gma/src/android/banner_view_internal_android.cc

```diff
@@ -142,15 +145,13 @@ BannerViewInternalAndroid::BannerViewInternalAndroid(BannerView* base)
 
 BannerViewInternalAndroid::~BannerViewInternalAndroid() {
   firebase::MutexLock lock(mutex_);
-  JNIEnv* env = ::firebase::gma::GetJNI();
-
-  // The application should have invoked Destroy already, but
-  // invoke it now just in case they haven't in the hope that
-  // we can prevent leaking memory.
-  env->CallVoidMethod(
-      helper_, banner_view_helper::GetMethodId(banner_view_helper::kDestroy),
-      /*callbackDataPtr=*/nullptr, /*destructor_invocation=*/true);
+  if (!destroyed_) {
+    LogWarning(
+        "Warning: AdView destructor invoked before the application "
+        " called Destroy() on the object. A Memory Leak may occur.");
```

Comment:
I think it's probably fine to just warn the developer, but as a way to make this more bulletproof, what about using the cleanup notifier to delegate the Destroy steps if the AdView destructor is called?

You could register object A to the cleanup notifier, with a deletion callback that performs the same thing on object A's contents as Destroy() would (but ignoring the result).

Reply:
I think the problem is the ignoring the result part.  The corruption seems to stem from removing the object reference after the asynchronous destruction operation is kicked off but before it completes. 

Maybe in Java I can switch which C++ function callback is invoked. Instead of using a boolean to denote if BannerViewHelper::Destroy was invoked by the Destructor, as it is now, I could pass the global reference to the BannerViewHelper.  If the global reference is NULL then the current C++ Destroy -> Java Destroy() -> C+++ future path is invoked.  If the global reference is non-null, then Java calls back into a different C++ static function with the Global Reference as a parameter.  The C++ static function then releases the global reference of the Java Object that time, and the C++ destructor doesn't even touch it.

I can try it out tomorrow.

Reply:
Or maybe just add the reference as an optional parameter and keep the boolean.

Reply:
Done. PTAL.